### ISSUE TITLE
Abort on assertion failure

### ISSUE THREAD

**Issue by yvt** [yvt/openspades#780]
It should result in a program termination (preferably with a core dump and helpful message) rather than throwing an exception for various reasons.

**Comment by theunknownxy** [comment]
Here are my thoughts about the network data validation issue: Receiving malformed data is not an logic error. For example an malicious server shouldn't be able to crash the client by triggering an assertion. The client should rather disconnect from the server, which means that normal error handling is more appropriate (using exceptions or error codes).

**Comment by NotAFile** [comment]
Network code is especially critical in that regard, since special strictness on some things e.g. out of range packet IDs and values might make expansion of the protocol harder.

**Comment by theunknownxy** [comment]
@NotAFile I don't agree, because it's not possible to distinguish between an broken packet and a recently defined packet. The client wouldn't know if the unknown or corrupt data is important. The server just shouldn't send new packets to an "old" client.

**Comment by NotAFile** [comment]
You're kind of right about that now, as OpenSpades now implements packet 60 extinfo that can be used to query capabilities of the client. But keep in mind that things such as the handshake, version and extinfo packets are only possible because voxlap AoS ignores unknown packets. Implementing packet 60 extinfo in the server was also made more difficult because OpenSpades did not really ignore it, which then required a hardcoded workaround for OS <=0.1.3 on the server side.
